File: mobile_app/lib/features/claims/data/claim_service.dart
import 'dart:io';

import 'package:dio/dio.dart';

import '../../../core/network/api_client.dart';
import '../domain/models/claim.dart';
import '../domain/models/claim_reason.dart';

class ClaimService {
  ClaimService({Dio? dio}) : _dio = dio ?? ApiClient.instance.dio;

  final Dio _dio;

  /// POST /claims — le userId n'est jamais envoyé par le client, il est
  /// résolu côté backend depuis le token (voir ClaimController.createClaim),
  /// même principe que POST /bookings.
  Future<Claim> createClaim({
    required ClaimReasonType reason,
    int? bookingId,
    required String message,
    Map<String, String>? details,
  }) async {
    final response = await _dio.post<Map<String, dynamic>>(
      '/claims',
      data: {
        'reason': reason.apiValue,
        if (bookingId != null) 'bookingId': bookingId,
        'message': message,
        if (details != null && details.isNotEmpty) 'details': details,
      },
    );
    return Claim.fromJson(response.data!);
  }

  /// Étape optionnelle après createClaim() : joint une preuve (image ou PDF) à une
  /// réclamation existante — propriété vérifiée côté backend (ClaimService.addAttachment).
  Future<Claim> addAttachment(int claimId, File file) async {
    final formData = FormData.fromMap({
      'file': await MultipartFile.fromFile(file.path),
    });
    final response = await _dio.post<Map<String, dynamic>>(
      '/claims/$claimId/attachment',
      data: formData,
    );
    return Claim.fromJson(response.data!);
  }

  Future<List<Claim>> getMyClaims(int userId) async {
    final response = await _dio.get<List<dynamic>>('/claims/mine/$userId');
    if (response.data == null) return [];
    return (response.data as List<dynamic>)
        .map((e) => Claim.fromJson(e as Map<String, dynamic>))
        .toList();
  }
}

File: frontend/src/app/core/services/claim/claim.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

/** Aligné sur l'enum backend ClaimReason. */
export type ClaimReason = 'REFUND_REQUEST' | 'CANCELLATION' | 'DELAY' | 'DRIVER_ISSUE' | 'LOST_ITEM' | 'OTHER';

/** Aligné sur ClaimStatus (backend). */
export type ClaimStatus = 'RECEIVED' | 'IN_PROGRESS' | 'RESOLVED' | 'REJECTED';

export const CLAIM_REASON_LABELS: Record<ClaimReason, string> = {
  REFUND_REQUEST: 'Demande de remboursement',
  CANCELLATION: 'Annulation',
  DELAY: 'Retard',
  DRIVER_ISSUE: 'Problème avec le chauffeur',
  LOST_ITEM: 'Objet perdu',
  OTHER: 'Autre',
};

export const CLAIM_STATUS_LABELS: Record<ClaimStatus, string> = {
  RECEIVED: 'Reçue',
  IN_PROGRESS: 'En cours de traitement',
  RESOLVED: 'Résolue',
  REJECTED: 'Rejetée',
};

/**
 * `true` si une réservation doit obligatoirement être associée à la réclamation — revalidé côté
 * serveur de toute façon (ClaimReason.requiresBooking()), répliqué ici juste pour l'UI (afficher/
 * masquer le sélecteur de réservation).
 */
export function claimReasonRequiresBooking(reason: ClaimReason): boolean {
  return reason === 'REFUND_REQUEST' || reason === 'CANCELLATION' || reason === 'DELAY' || reason === 'DRIVER_ISSUE';
}

export interface CreateClaimRequest {
  reason: ClaimReason;
  bookingId?: number | null;
  message: string;
  details?: Record<string, string>;
}

/**
 * Aligné sur PassengerClaimResponse (backend) — jamais adminNote, réservé à la vue admin.
 * `booking` = BookingSummaryResponse, champs exacts non garantis ici : accès défensif côté UI.
 */
export interface PassengerClaim {
  id: number;
  reason: ClaimReason;
  status: ClaimStatus;
  booking: ({ id: number } & Record<string, unknown>) | null;
  message: string;
  details: Record<string, string> | null;
  resolutionMessage: string | null;
  createdAt: string;
  resolvedAt: string | null;
  attachmentPath: string | null;
  attachmentOriginalName: string | null;
  attachmentContentType: string | null;
}

@Injectable({ providedIn: 'root' })
export class ClaimService {
  private http = inject(HttpClient);

  create(body: CreateClaimRequest): Observable<PassengerClaim> {
    return this.http.post<PassengerClaim>('/claims', body);
  }

  listMine(userId: number): Observable<PassengerClaim[]> {
    return this.http.get<PassengerClaim[]>(`/claims/mine/${userId}`);
  }

  /** Étape optionnelle après create() : joint une preuve (image ou PDF) à une réclamation existante. */
  addAttachment(claimId: number, file: File): Observable<PassengerClaim> {
    const form = new FormData();
    form.append('file', file);
    return this.http.post<PassengerClaim>(`/claims/${claimId}/attachment`, form);
  }
}
